comp/trace/agent/agent.go

```diff
@@ -0,0 +1,154 @@
+	fx.In
+
+	Lc         fx.Lifecycle
+	Shutdowner fx.Shutdowner
+
+	Params *Params
+	Config config.Component
+}
+
+type agent struct {
+	*pkgagent.Agent
+
+	cancel             context.CancelFunc
+	config             config.Component
+	ctx                context.Context
+	params             *Params
+	shutter            fx.Shutdowner
+	telemetryCollector telemetry.TelemetryCollector
+}
+
+func newAgent(deps dependencies) Component {
+	telemetryCollector := telemetry.NewCollector(deps.Config.Object())
+
+	// Several related non-components require a shared context to gracefully stop.
+	ctx, cancel := context.WithCancel(context.Background())
+	ag := &agent{
+		Agent: pkgagent.NewAgent(
+			ctx,
+			deps.Config.Object(),
+			telemetryCollector,
+		),
+		cancel:             cancel,
+		config:             deps.Config,
+		params:             deps.Params,
+		shutter:            deps.Shutdowner,
+		telemetryCollector: telemetryCollector,
+	}
+
+	deps.Lc.Append(fx.Hook{OnStart: ag.start, OnStop: ag.stop})
+
+	return ag
+}
+
+// Provided ctx has a timeout, so it can't be used for gracefully stopping long-running components.
+// This context is cancelled on a deadline, so it would stop the agent after starting it.
+func (ag *agent) start(_ context.Context) error {
+	// Handle stops properly
+	go func() {
+		defer watchdog.LogOnPanic()
+		handleSignal(ag.shutter)
+	}()
+
+	if ag.params.CPUProfile != "" {
+		f, err := os.Create(ag.params.CPUProfile)
+		if err != nil {
+			log.Error(err)
+		}
+		pprof.StartCPUProfile(f) //nolint:errcheck
+		log.Info("CPU profiling started...")
+		defer pprof.StopCPUProfile()
+	}
+
+	if ag.params.PIDFilePath != "" {
+		err := pidfile.WritePID(ag.params.PIDFilePath)
+		if err != nil {
+			ag.telemetryCollector.SendStartupError(telemetry.CantWritePIDFile, err)
+			log.Criticalf("Error writing PID file, exiting: %v", err)
+			os.Exit(1)
+		}
+
+		log.Infof("PID '%d' written to PID file '%s'", os.Getpid(), ag.params.PIDFilePath)
+		defer os.Remove(ag.params.PIDFilePath)
+	}
+
+	traceCfg := ag.config.Object()
+	apiConfigHandler := ag.config.SetHandler()
+
+	if err := runAgentSidekicks(ag.ctx, traceCfg, apiConfigHandler, ag.telemetryCollector); err != nil {
+		return err
+	}
+
+	go ag.Run()
+
+	return nil
+}
+
+func (ag *agent) stop(_ context.Context) error {
+	ag.cancel()
+
+	// collect memory profile
+	if ag.params.MemProfile != "" {
+		f, err := os.Create(ag.params.MemProfile)
+		if err != nil {
+			log.Error("Could not create memory profile: ", err)
+		}
+
+		// get up-to-date statistics
+		runtime.GC()
+		// Not using WriteHeapProfile but instead calling WriteTo to
+		// make sure we pass debug=1 and resolve pointers to names.
+		if err := pprof.Lookup("heap").WriteTo(f, 1); err != nil {
+			log.Error("Could not write memory profile: ", err)
+		}
+		f.Close()
+	}
+
+	return nil
+}
+
+// handleSignal closes a channel to exit cleanly from routines
+func handleSignal(shutter fx.Shutdowner) {
+	sigChan := make(chan os.Signal, 10)
```

Comment:
why `10`?

Reply:
@katiehockman That was actually like this in the original code that I moved from `pkg/trace/agent`. I left it as is.

Reply:
Curious: almost everywhere we use buffered channels with 1 element length, except here and some code in `pkg/proces/util` (where the channel has a length of 100 elements).

According to `os/signal` docs:

> Package signal will not block sending to c: the caller must ensure that c has sufficient buffer space to keep up with the expected signal rate. For a channel used for notification of just one signal value, a buffer of size 1 is sufficient.

There seems not to be a good reason to create a buffered channel with length bigger than 1, so I fixed it.

Reply:
🙏  Thanks